Flexbox solution fails unit tests - tech docs project

The testing for this doesn't allow for flexbox configuration, since the contents of the flexbox need to be wrapped in a div with position set to fixed for it to work, and this causes the following conditions to fail:

CONTENT
 9. The first element within the navbar should be a <header> which contains text that describes the topic of the technical documentation.
(failed because first element was the div mentioned above)

LAYOUT 
1. On regular sized devices (laptops, desktops), the element with id="navbar" should be shown on the left side of the screen and should always be visible to the user.
failed with "AssertionError: expected -1000 to be close to 0 +/- 10"

If your intention is to get people to use another approach, like fixed and absolute positioning for the two main sections - kindly ignore this issue!

#### Browser Information
Using iMac w/Chrome browser Version 56.0.2924.87 (64-bit)


#### Your Code / Link to Your Pen
http://codepen.io/sue888/pen/OpRoqB

@QuincyLarson Before going ahead with creating a PR for this issue I would like to better understand the purpose of the "technical-documentation-page" assignment. Looking at the [existing tests](https://github.com/freeCodeCamp/testable-projects-fcc/blob/master/src/project-tests/technical-docs-tests.js), I am guessing the goal for the assignment is to have the student use the `<nav>` (and related) elements along with some other basic elements like `<code>`, `<li>`, and `<p>`. 

Concerning the layout, the `<nav>` element must stay visible, and the page should use at least one media query.

If the above is correct and I am not missing anything, then I recommend the following approach to the 2 issues described by @sue888.

**The header requirement**

"The first element within the navbar should be a `<header>` which contains text that describes the topic." 

This can be improved by making the requirement slightly more flexible: 

"The navbar element should contain one `<header>` element which contains text that describes the topic". 

The existing test checks for `firstElementChild` which does not allow any container `<divs>` because the `<header>` must be the very first and top child. I would change this to use `getElementsByTagName('header').length === 1`. `getElementsByTagName()` searches the entire sub-tree so this would allow for any number of containers outside the `<header>`. I think this test alone would be good enough, but if we wanted to make sure the header appears before any navbar links, I could try to add a test for that too. Something like "The `<header>` element inside the navbar must appear before any `<a>` elements with the class of "nav-link"."

**NOTE:** The codepen provided by @sue888 uses the `<header>` element for *every* link within the navbar, so my new proposed test would still cause that code to fail. However, I believe it should fail because using the `<header>` element for each `<a>` link in the navbar is not recommended. The `<header>` element should only be used once within the navbar.

**The layout requirement**

"On regular sized devices (laptops, desktops), the element with id="navbar" should be shown on the left side of the screen and should always be visible to the user."
failed with "AssertionError: expected -1000 to be close to 0 +/- 10

I believe this is working as intended. If you look at the [working example pen](https://codepen.io/freeCodeCamp/pen/NdrKKL), the element with id="navbar" has CSS specifying   `position:fixed`, keeping the `<nav>` element on the top left.

But on the [CodePen provided by @sue888](https://codepen.io/sue888/pen/OpRoqB), the element with id="navbar" does not have a fixed position.

The navbar *appears* to stay at a fixed position because a container within the navbar (`<div id="fixed-or-not">`) does have a fixed position. Visually all of the navbar content stays on the top left, but in reality the `<nav>` element actually does scroll off the top of the viewport when the user scrolls down. 

I have not tried it, and I think it would be out of scope for me, but I believe it would be possible to make the navbar a fixed position while still using a flexbox layout. So I don't see a problem with keeping this test as-is, and having students make sure the `<nav>` element itself stays on the top left. I.e. this test is working as intended and the CodePen provided by @sue888 should be fixed instead.

Your thoughts?

--- a/build/bundle.js
+++ b/build/bundle.js
@@ -20541,11 +20541,11 @@ var FCC_Global =
 	        FCC_Global.assert.strictEqual(firstChildHeaderQty.length, classQty, "Not all elements with the class 'main-section' have a <header> element as a first element child ");
 	    };
 
-	    describe("Technical Documentation Page tests", function () {
+	    describe('Technical Documentation Page tests', function () {
 	        describe('#Content', function () {
 	            it('1. I can see a <main> element with a corresponding id="main-doc", which contains the ' + 'page\'s main content (technical documentation).', function () {
 	                FCC_Global.assert.isNotNull(document.getElementById('main-doc'), "There is no element with an id of 'main-doc' ");
-	                FCC_Global.assert.strictEqual(document.getElementById('main-doc').nodeName, "MAIN", "The 'main-doc' element should be a <main> ");
+	                FCC_Global.assert.strictEqual(document.getElementById('main-doc').nodeName, 'MAIN', "The 'main-doc' element should be a <main> ");
 	            });
 
 	            it('2. Within the #main-doc ( <main> ) element, I can see several <section> elements, each with a class ' + 'of "main-section". There should be a minimum of 5.', function () {
@@ -20578,7 +20578,7 @@ var FCC_Global =
 
 	                var headerText = mainSections.map(function (el, i) {
 	                    FCC_Global.assert.isAbove(el.firstElementChild.innerText.length, 0, 'All headers must contain text ');
-	                    //remove leading / trailing spaces  in headerText and then replace all other spaces with underscores.
+	                    // remove leading / trailing spaces  in headerText and then replace all other spaces with underscores.
 	                    return el.firstElementChild.innerText.trim().replace(/\s/g, '_');
 	                });
 
@@ -20594,24 +20594,24 @@ var FCC_Global =
 	            });
 
 	            it('5. The .main-section elements should contain at least 10 <p> elements total (not each).', function () {
-	                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section p').length, 10, "There are not at least 10 <p> " + "elements throughout all of the elements with the class of 'main-section' ");
+	                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section p').length, 10, 'There are not at least 10 <p> ' + "elements throughout all of the elements with the class of 'main-section' ");
 	            });
 
 	            it('6. The .main-section elements should contain at least 5 <code> elements total (not each).', function () {
-	                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section code').length, 5, "There are not at least 5 <code> " + "elements throughout all of the elements with the class of 'main-section' ");
+	                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section code').length, 5, 'There are not at least 5 <code> ' + "elements throughout all of the elements with the class of 'main-section' ");
 	            });
 
 	            it('7. The .main-section elements should contain at least 5 <li> items total (not each).', function () {
-	                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section li').length, 5, "There are not " + "at least 5 <li> elements throughout all of the elements with the class of 'main-section' ");
+	                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section li').length, 5, 'There are not ' + "at least 5 <li> elements throughout all of the elements with the class of 'main-section' ");
 	            });
 
 	            it('8. I can see a <nav> element with a corresponding id="navbar".', function () {
 	                FCC_Global.assert.isNotNull(document.getElementById('navbar'), "There is no element with the id of 'navbar'.");
-	                FCC_Global.assert.strictEqual(document.getElementById('navbar').nodeName, "NAV", 'The element with the id of ' + '"navbar" is not a <nav> element ');
+	                FCC_Global.assert.strictEqual(document.getElementById('navbar').nodeName, 'NAV', 'The element with the id of ' + '"navbar" is not a <nav> element ');
 	            });
 
-	            it('9. The first element within the navbar should be a <header> which contains text that describes the topic ' + 'of the technical documentation.', function () {
-	                FCC_Global.assert.strictEqual(document.getElementById('navbar').firstElementChild.nodeName, 'HEADER', 'The first ' + 'element within #navbar should be a <header> ');
+	            it('9. The navbar element should contain one <header> element which contains text that describes the topic ' + 'of the technical documentation.', function () {
+	                FCC_Global.assert.strictEqual(document.getElementById('navbar').getElementsByTagName('header').length, 1, 'The navbar ' + 'element should contain only one <header> element.');
 	            });
 
 	            it('10. Additionally, the navbar should contain link (<a>) elements with the class of "nav-link". There ' + 'should be one for every element with the class "main-section".', function () {
@@ -20625,7 +20625,21 @@ var FCC_Global =
 	                FCC_Global.assert.strictEqual(typeQty.length, navLinkQty, 'Not all of the elements with the class of "nav-link" are ' + '<a> elements ');
 	            });
 
-	            it('11. Each element with the class of "nav-link" should contain text that corresponds to the <header> text ' + 'within each <section> (e.g. if you have a "Hello world" section/header, your navbar should have an ' + 'element which contains the text "Hello world").', function () {
+	            it('11. The <header> element in the navbar must come before any link (<a>) elements in the navbar.', function () {
+	                var navLinks = document.querySelectorAll('#navbar a.nav-link');
+	                var header = document.querySelectorAll('#navbar header').item(0);
+	                var allAbove = true;
+
+	                navLinks.forEach(function (navLink) {
+	                    if (!(header.compareDocumentPosition(navLink) & Node.DOCUMENT_POSITION_FOLLOWING)) {
+	                        allAbove = false;
+	                    }
+	                });
+
+	                FCC_Global.assert.strictEqual(allAbove, true, 'The <header> element contained by the navbar must be come before ' + 'any link (<a>) elements contained by the navbar');
+	            });
+
+	            it('12. Each element with the class of "nav-link" should contain text that corresponds to the <header> text ' + 'within each <section> (e.g. if you have a "Hello world" section/header, your navbar should have an ' + 'element which contains the text "Hello world").', function () {
 	                FCC_Global.assert.isAbove(classArray('nav-link').length, 0, 'No elements with the class "nav-link" have been defined ');
 	                var headerText = classArray('main-section').map(function (el, i) {
 	                    return el.firstElementChild.innerText.trim();
@@ -20641,7 +20655,7 @@ var FCC_Global =
 	                FCC_Global.assert.strictEqual(remainder.length, 0, 'Check that these headers have corresponding .nav-link elements and ' + 'be mindful of case! : ' + remainder + ' ');
 	            });
 
-	            it('12. When I click on a navbar element, the page should navigate to the corresponding section of the ' + '"main-doc" element (e.g. If I click on a "nav-link" element that contains the text "Hello world", the page ' + 'navigates to a <section> element that has that id and contains the corresponding <header>.', function () {
+	            it('13. When I click on a navbar element, the page should navigate to the corresponding section of the ' + '"main-doc" element (e.g. If I click on a "nav-link" element that contains the text "Hello world", the page ' + 'navigates to a <section> element that has that id and contains the corresponding <header>.', function () {
 	                var navLinkQty = document.getElementsByClassName('nav-link').length;
 	                FCC_Global.assert.isAbove(navLinkQty, 0, 'No elements with the class "nav-link" have been defined ');
 	                var hasHref = classArray('nav-link').filter(function (el) {
@@ -20666,11 +20680,11 @@ var FCC_Global =
 	                var windowWidth = document.documentElement.clientWidth || window.innerWidth;
 	                FCC_Global.assert.isAbove(windowWidth, 850, 'Please run this test in a larger window (before any media queries) ');
 	                var navbar = document.getElementById('navbar');
-	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10);
-	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10);
+	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10, 'Left of bounding rectangle is not correct.');
+	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10, 'Top of bounding rectangle is not correct.');
 	                window.scroll(0, 1000);
-	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10);
-	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10);
+	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10, 'After scroll: Left of bounding rectangle is not correct.');
+	                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10, 'After scroll: Top of bounding rectangle is not correct.');
 	                window.scroll(0, 0);
 	            });
 

--- a/src/project-tests/technical-docs-tests.js
+++ b/src/project-tests/technical-docs-tests.js
@@ -3,19 +3,19 @@ export default function createTechnicalDocsPageTests() {
     const classArray = (className) => {
         return Array.from(document.getElementsByClassName(className));
     };
-    
+
     const sharedHeaderTest = (classQty) => {
       let firstChildHeaderQty = classArray('main-section').filter(el => el.firstElementChild.nodeName === 'HEADER');
       FCC_Global.assert.strictEqual(firstChildHeaderQty.length, classQty, "Not all elements with the class 'main-section' have a <header> element as a first element child ");
-    }
+    };
 
-    describe("Technical Documentation Page tests", function() {
+    describe('Technical Documentation Page tests', function() {
         describe('#Content', function() {
             it('1. I can see a <main> element with a corresponding id="main-doc", which contains the ' +
                 'page\'s main content (technical documentation).',
                 function() {
                     FCC_Global.assert.isNotNull(document.getElementById('main-doc'), "There is no element with an id of 'main-doc' ");
-                    FCC_Global.assert.strictEqual(document.getElementById('main-doc').nodeName, "MAIN",
+                    FCC_Global.assert.strictEqual(document.getElementById('main-doc').nodeName, 'MAIN',
                         "The 'main-doc' element should be a <main> ");
                 });
 
@@ -51,53 +51,53 @@ export default function createTechnicalDocsPageTests() {
                 '<section> that contains the header "Javascript and Java" should have a corresponding ' +
                 'id="Javascript_and_Java").',
                 function() {
-                  
+
                     const mainSections = classArray('main-section');
                     FCC_Global.assert.isAbove(mainSections.length, 0, 'No elements with the class "main-section" are defined ');
-                    
+
                     sharedHeaderTest(mainSections.length);
-                    
+
                     const headerText = mainSections.map((el, i) => {
-                      FCC_Global.assert.isAbove(el.firstElementChild.innerText.length, 0, 'All headers must contain text ')
-                      //remove leading / trailing spaces  in headerText and then replace all other spaces with underscores.
+                      FCC_Global.assert.isAbove(el.firstElementChild.innerText.length, 0, 'All headers must contain text ');
+                      // remove leading / trailing spaces  in headerText and then replace all other spaces with underscores.
                       return el.firstElementChild.innerText.trim().replace(/\s/g, '_');
                     });
-                    
+
                     const mainSectionIDs = mainSections.map((el, i) => {
-                      FCC_Global.assert.strictEqual(el.hasAttribute('id'), true, "Each 'main-section' should have an id attribute ")
+                      FCC_Global.assert.strictEqual(el.hasAttribute('id'), true, "Each 'main-section' should have an id attribute ");
                       return el.id;
                     });
-                    
+
                     const remainder = headerText.filter(str => mainSectionIDs.indexOf(str) === -1);
                     FCC_Global.assert.strictEqual(remainder.length, 0, 'Some "main-section" elements are missing the following ids (don\'t forget to replace spaces with underscores!) : ' + remainder + ' ');
                 });
 
             it('5. The .main-section elements should contain at least 10 <p> elements total (not each).', function() {
-                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section p').length, 10, "There are not at least 10 <p> " +
+                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section p').length, 10, 'There are not at least 10 <p> ' +
                     "elements throughout all of the elements with the class of 'main-section' ");
             });
 
             it('6. The .main-section elements should contain at least 5 <code> elements total (not each).', function() {
-                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section code').length, 5, "There are not at least 5 <code> " +
+                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section code').length, 5, 'There are not at least 5 <code> ' +
                     "elements throughout all of the elements with the class of 'main-section' ");
             });
 
             it('7. The .main-section elements should contain at least 5 <li> items total (not each).', function() {
-                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section li').length, 5, "There are not " +
+                FCC_Global.assert.isAtLeast(document.querySelectorAll('.main-section li').length, 5, 'There are not ' +
                     "at least 5 <li> elements throughout all of the elements with the class of 'main-section' ");
             });
 
             it('8. I can see a <nav> element with a corresponding id="navbar".', function() {
                 FCC_Global.assert.isNotNull(document.getElementById('navbar'), "There is no element with the id of 'navbar'.");
-                FCC_Global.assert.strictEqual(document.getElementById('navbar').nodeName, "NAV", 'The element with the id of ' +
+                FCC_Global.assert.strictEqual(document.getElementById('navbar').nodeName, 'NAV', 'The element with the id of ' +
                     '"navbar" is not a <nav> element ');
             });
 
-            it('9. The first element within the navbar should be a <header> which contains text that describes the topic ' +
+            it('9. The navbar element should contain one <header> element which contains text that describes the topic ' +
                 'of the technical documentation.',
                 function() {
-                    FCC_Global.assert.strictEqual(document.getElementById('navbar').firstElementChild.nodeName, 'HEADER', 'The first ' +
-                        'element within #navbar should be a <header> ');
+                    FCC_Global.assert.strictEqual(document.getElementById('navbar').getElementsByTagName('header').length, 1, 'The navbar ' +
+                        'element should contain only one <header> element.');
                 });
 
             it('10. Additionally, the navbar should contain link (<a>) elements with the class of "nav-link". There ' +
@@ -114,20 +114,36 @@ export default function createTechnicalDocsPageTests() {
                         '<a> elements ');
                 });
 
-            it('11. Each element with the class of "nav-link" should contain text that corresponds to the <header> text ' +
+            it('11. The <header> element in the navbar must come before any link (<a>) elements in the navbar.',
+                function() {
+                  const navLinks = document.querySelectorAll('#navbar a.nav-link');
+                  const header = document.querySelectorAll('#navbar header').item(0);
+                  let allAbove = true;
+
+                  navLinks.forEach(navLink => {
+                    if (!(header.compareDocumentPosition(navLink) & Node.DOCUMENT_POSITION_FOLLOWING)) {
+                      allAbove = false;
+                    }
+                  });
+
+                  FCC_Global.assert.strictEqual(allAbove, true, 'The <header> element contained by the navbar must be come before ' +
+                      'any link (<a>) elements contained by the navbar');
+                });
+
+            it('12. Each element with the class of "nav-link" should contain text that corresponds to the <header> text ' +
                 'within each <section> (e.g. if you have a "Hello world" section/header, your navbar should have an ' +
                 'element which contains the text "Hello world").',
                 function() {
                     FCC_Global.assert.isAbove(classArray('nav-link').length, 0, 'No elements with the class "nav-link" have been defined ');
                     const headerText = classArray('main-section').map((el, i) => el.firstElementChild.innerText.trim());
-                    const linkText = classArray('nav-link').map((el, i) => /[^\n\t\f\r\v]+/.exec(el.innerText)[0]);
+                    const linkText = classArray('nav-link').map((el, i) => (/[^\n\t\f\r\v]+/).exec(el.innerText)[0]);
                     // use indexOf instead of matching index for index, in case for some reason they have them out of order
                     const remainder = headerText.filter(str => linkText.indexOf(str) === -1);
                     FCC_Global.assert.strictEqual(remainder.length, 0, 'Check that these headers have corresponding .nav-link elements and ' +
                         'be mindful of case! : ' + remainder + ' ');
                 });
 
-            it('12. When I click on a navbar element, the page should navigate to the corresponding section of the ' +
+            it('13. When I click on a navbar element, the page should navigate to the corresponding section of the ' +
                 '"main-doc" element (e.g. If I click on a "nav-link" element that contains the text "Hello world", the page ' +
                 'navigates to a <section> element that has that id and contains the corresponding <header>.',
                 function() {
@@ -136,7 +152,7 @@ export default function createTechnicalDocsPageTests() {
                     const hasHref = classArray('nav-link').filter(el => el.hasAttribute('href'));
                     const hrefValues = classArray('nav-link').map((el, i) => el.getAttribute('href'));
                     const mainSectionIDs = classArray('main-section').map((el, i) => el.id);
-                    const missingHrefValues = mainSectionIDs.filter(str => hrefValues.indexOf('#' + str) === -1)
+                    const missingHrefValues = mainSectionIDs.filter(str => hrefValues.indexOf('#' + str) === -1);
                     FCC_Global.assert.strictEqual(hasHref.length, navLinkQty, 'Every .nav-link does not have an href attribute.');
                     FCC_Global.assert.strictEqual(missingHrefValues.length, 0, 'Every .nav-link should have an href value ' +
                         'that links it to its corresponding .main-section (e.g. href="#Introduction"). Check that these ' +
@@ -147,13 +163,13 @@ export default function createTechnicalDocsPageTests() {
         describe('#Layout', function() {
             it('1. On regular sized devices (laptops, desktops), the element with id="navbar" should be shown on the left side of the screen and should always be visible to the user.', function() {
                 const windowWidth = document.documentElement.clientWidth || window.innerWidth;
-                FCC_Global.assert.isAbove(windowWidth, 850, 'Please run this test in a larger window (before any media queries) ')
+                FCC_Global.assert.isAbove(windowWidth, 850, 'Please run this test in a larger window (before any media queries) ');
                 var navbar = document.getElementById('navbar');
-                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10);
-                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10);
+                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10, 'Left of bounding rectangle is not correct.');
+                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10, 'Top of bounding rectangle is not correct.');
                 window.scroll(0, 1000);
-                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10);
-                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10);
+                FCC_Global.assert.approximately(navbar.getBoundingClientRect().left, 0, 10, 'After scroll: Left of bounding rectangle is not correct.');
+                FCC_Global.assert.approximately(navbar.getBoundingClientRect().top, 0, 10, 'After scroll: Top of bounding rectangle is not correct.');
                 window.scroll(0, 0);
             });
 

